nx/lib/nx/random.ex

```diff
@@ -0,0 +1,374 @@
+
+    new_ks = Nx.stack([k2, k3, k1])
+    new_rs = Nx.stack([r2, r1])
+
+    {new_xs, new_ks, new_rs}
+  end
+
+  defnp rotate_left(x, rot) do
+    x <<< rot ||| x >>> (@nbits - rot)
+  end
+
+  defp float_info(type) do
+    case type do
+      {:bf, 16} -> [exp: 8, mantissa: 7]
+      {:f, 16} -> [exp: 5, mantissa: 11]
+      {:f, 32} -> [exp: 8, mantissa: 23]
+      {:f, 64} -> [exp: 11, mantissa: 52]
+    end
+  end
+
+  @doc """
+  Sample uniform random integer values in [min_val, max_val).
+
+  ## Options
+
+    * `:type` - an int type for the returned tensor
+
+    * `:shape` - shape of the returned tensor
+
+  """
+  defn randint(key, min_val, max_val, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:s, 32})
+    assert_key(key)
+
+    shape = opts[:shape]
+    type = {_, nbits} = opts[:type]
+    case type do
+      {:u, _} -> :ok
+      {:s, _} -> :ok
+      _ -> raise ArgumentError,
+            "expected integer type, got type #{inspect(type)}"
+    end
+
+    keys = split(key)
+
+    min_val = Nx.broadcast(min_val, shape)
+    max_val = Nx.broadcast(max_val, shape)
+
+    higher_bits = random_bits(keys[0], shape: shape)
+    lower_bits = random_bits(keys[1], shape: shape)
+    span = max_val - min_val
+
+    multiplier =
+      Nx.power(2, Nx.quotient(nbits, 2))
+      |> Nx.remainder(span)
+      |> Nx.power(2)
+      |> Nx.remainder(span)
+
+    offset =
+      higher_bits
+      |> Nx.remainder(span)
+      |> Nx.multiply(multiplier)
+      |> Nx.add(Nx.remainder(lower_bits, span))
+      |> Nx.remainder(span)
+
+    Nx.as_type(min_val + offset, type)
+  end
+
+  @doc """
+  Sample uniform float values in [min_val, max_val).
+
+  ## Options
+
+    * `:type` - an float type for the returned tensor
+
+    * `:shape` - shape of the returned tensor
+
+    * `:min_val` - minimum value, default is 0.0
+
+    * `:max_val` - maximum value, default is 1.0
+
+  """
+  defn uniform(key, opts \\ []) do
+    opts = keyword!(opts, shape: {1}, type: {:f, 32}, min_val: 0.0, max_val: 1.0)
+
+    assert_key(key)
+
+    type = opts[:type]
+    case type do
+      {:f, _} -> :ok
+
+      _ -> raise ArgumentError,
+            "expected float type, got type #{inspect(type)}"
+    end
+
+    info = transform(opts[:type], &float_info(&1))
+
+    shape = opts[:shape]
+    type = {_dtype, nbits} = opts[:type]
+    min_val = opts[:min_val]
+    max_val = opts[:max_val]
+
+    u_one = Nx.tensor(1065353216, type: {:u, nbits})
+    #Equivalent of constant - Nx.tensor(1.0, type: type) |> Nx.bitcast({:u, nbits})
+
+    random_bits(key, shape: shape)
+    |> Nx.as_type({:u, nbits})
+    |> Nx.right_shift(Nx.tensor(nbits - info[:mantissa], type: {:u, nbits}))
+    |> Nx.bitwise_or(u_one)
+    |> bitcast(type)
+    |> Nx.subtract(Nx.tensor(1.0, type: type))
+    |> Nx.multiply(max_val - min_val)
+    |> Nx.add(min_val)
+    |> Nx.reshape(shape)
+    |> Nx.max(min_val)
+  end
+
+  deftransformp bitcast(tensor, type), do: Nx.bitcast(tensor, type)

```

Comment:
why does this need to be deftransform?

Reply:
> We can't have bitcast in Torchx.
>
> Since it's the beggining of the function, I would rather do the bitcast in a transform in pure elixir if it is indeed necessary.

I followed that suggestion and made it deftransform because transform() is getting deprecated

Reply:
Yeah, the point is that you can use `Nx.bitcast` directly in `defn`, doesn't need to be a transform :)

Reply:
> Yeah, the point is that you can use `Nx.bitcast` directly in `defn`, doesn't need to be a transform :)

Sorry, but now I'm confused on relationship between defn and torchx. Should we use another solution if we can't have bitcast in Torchx or what exactly is the problem? I know we can use Nx.bitcast in defn and it was like that at the beginning but I was refering to @polvalente suggestion.

Reply:
Ah, I see the other comment.

Wrapping it in a `deftransform` doesn't change anything. `deftransform` allows you to freely call arbitrary functions `Enum.*` or use constructs like `for`, however calling functions from `Nx.*` works the same way, it builds up the numerical expression.

So the point was that ideally we would avoid `Nx.bitcast` altogether. So given

```elixir
u_one = Nx.tensor(1.0, type: type) |> Nx.bitcast({:u, nbits})
```

we know all the operands when building the expression (`1.0`, `type`, `nbits`), so we could compute `u_one` in pure elixir, without `Nx.bitcast`.

But I see we still use `bitcast` in the other place and I'm not sure if we can easily eliminate it, because it operates on the generated tensor.

If this is not clear or there's anything about `defn`/transform that confuses you let me know!